### PR TITLE
Application Keys Import/Export

### PR DESCRIPTION
**Pull request by kasvith**
In following PR
- Added support for Application Import/Export
- Added support for updating subscribers for an application(will leave already subscribed subscribers, new ones will be added) only on `--update` flag

>  `apimcli` was configured to support this feature

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ExportApiServiceImpl.java [hunk 1]
```diff
@@ -90,27 +90,24 @@ public Response exportApplicationsGet(String appName, String appOwner, Boolean w
                 return Response.status(Response.Status.FORBIDDEN).entity(errorMsg).build();
             }
 
+            // clear all duplicate keys with tokens
+            applicationDetails.getKeys().clear();
+
```
Comment (Rajith90) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ExportApiServiceImpl.java]: Check for the possibilities of NPE if applicationDetails.getKeys() returns null

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ImportApiServiceImpl.java [hunk 2]
```diff
@@ -89,13 +89,17 @@ public Response importApplicationsPost(InputStream fileInputStream, Attachment f
             // decode Oauth secrets
             OAuthApplicationInfo productionOAuthApplicationInfo = applicationDetails.getOAuthApp("PRODUCTION");
             if (productionOAuthApplicationInfo != null) {
-                String decodedConsumerSecretBytes = productionOAuthApplicationInfo.getClientSecret();
-                productionOAuthApplicationInfo.setClientSecret(new String(Base64.decodeBase64(decodedConsumerSecretBytes)));
+                String encodedConsumerSecretBytes = productionOAuthApplicationInfo.getClientSecret();
+                String decodedConsumerSecret = new String(Base64.decodeBase64(encodedConsumerSecretBytes));
+                productionOAuthApplicationInfo.setClientSecret(decodedConsumerSecret);
+                applicationDetails.addKey(getAPIKeyFromOauthApp("PRODUCTION", productionOAuthApplicationInfo));
```
Comment (Rajith90) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ImportApiServiceImpl.java]: Use constants for "Production"

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ImportApiServiceImpl.java [hunk 3]
```diff
@@ -89,13 +89,17 @@ public Response importApplicationsPost(InputStream fileInputStream, Attachment f
...
             }
             OAuthApplicationInfo sandboxOAuthApplicationInfo = applicationDetails.getOAuthApp("SANDBOX");
             if (sandboxOAuthApplicationInfo != null) {
-                String decodedConsumerSecretBytes = sandboxOAuthApplicationInfo.getClientSecret();
-                sandboxOAuthApplicationInfo.setClientSecret(new String(Base64.decodeBase64(decodedConsumerSecretBytes)));
+                String encodedConsumerSecretBytes = sandboxOAuthApplicationInfo.getClientSecret();
+                String decodedConsumerSecret = new String(Base64.decodeBase64(encodedConsumerSecretBytes));
+                sandboxOAuthApplicationInfo.setClientSecret(decodedConsumerSecret);
+                applicationDetails.addKey(getAPIKeyFromOauthApp("SANDBOX", sandboxOAuthApplicationInfo));
```
Comment (Rajith90) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ImportApiServiceImpl.java]: Use constants for "SANDBOX"

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ImportApiServiceImpl.java [hunk 4]
```diff
@@ -164,6 +166,26 @@ public Response importApplicationsPost(InputStream fileInputStream, Attachment f
...
+     * @return An APIKey containing keys from OAuthApplication
+     */
+    private APIKey getAPIKeyFromOauthApp(String type, OAuthApplicationInfo oAuthApplicationInfo){
+        APIKey apiKey = new APIKey();
+        apiKey.setType(type);
+        apiKey.setConsumerKey(oAuthApplicationInfo.getClientId());
+        apiKey.setConsumerSecret(oAuthApplicationInfo.getClientSecret());
+        apiKey.setGrantTypes((String) oAuthApplicationInfo.getParameter("grant_types"));
+        if (apiKey.getGrantTypes().contains("implicit") && apiKey.getGrantTypes().contains("code")){
+            apiKey.setCallbackUrl((String) oAuthApplicationInfo.getParameter("redirect_uris"));
```
Comment (Rajith90) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ImportApiServiceImpl.java]: Use constants

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ImportApiServiceImpl.java [hunk 5]
```diff
@@ -164,6 +166,26 @@ public Response importApplicationsPost(InputStream fileInputStream, Attachment f
...
+        }
+        apiKey.setValidityPeriod(3600);
+        apiKey.setTokenScope("am_application_scope default");
+        return apiKey;
```
Comment (Rajith90) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ImportApiServiceImpl.java]: Do we have to hard code these values

Comment (kasvith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ImportApiServiceImpl.java]: As an offline discussion with @harsha89 it was decided to put this value 3600

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ExportApiServiceImpl.java (hunk 1 above)
Comment (kasvith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ExportApiServiceImpl.java]: Keys are initialized in the `Application` class.

**Pull request merged**